src/Network/MQTT/Encoder.hs

```diff
@@ -0,0 +1,166 @@
+              , toBit connectCleanSession                      `shiftL` 1
+              ]
+
+encodePublishVariableHeader :: PublishPacket -> Builder
+encodePublishVariableHeader PublishPacket{..} =
+  encodeText (unTopic $ messageTopic publishMessage)
+  <> maybe mempty encodePacketIdentifier publishPacketIdentifier
+
+encodeText :: Text -> Builder
+encodeText text = word16BE (fromIntegral (BS.length encodedUtf)) <> byteString encodedUtf
+  where
+    encodedUtf = encodeUtf8 text
+
+encodeBytes :: BS.ByteString -> Builder
+encodeBytes bytes = word16BE (fromIntegral (BS.length bytes)) <> byteString bytes
+
+encodeConnackVariableHeader :: ConnackPacket -> Builder
+encodeConnackVariableHeader ConnackPacket{..} = word8 (toBit connackSessionPresent)
+                                              <> word8 (returnCodeValue connackReturnCode)
+
+returnCodeValue :: ConnackReturnCode -> Word8
+returnCodeValue Accepted              = 0
+returnCodeValue UnacceptableProtocol  = 1
+returnCodeValue IdentifierRejected    = 2
+returnCodeValue ServerUnavailable     = 3
+returnCodeValue BadUserNameOrPassword = 4
+returnCodeValue NotAuthorized         = 5
+
+encodePacketIdentifier :: PacketIdentifier -> Builder
+encodePacketIdentifier PacketIdentifier{..} = word16BE unPacketIdentifier
+
+payload :: Packet -> Builder
+payload (CONNECT connectPacket)         = encodeConnectPayload connectPacket
+payload (PUBLISH publishPacket)         = encodePublishPayload publishPacket
+payload (SUBSCRIBE subscribePacket)     = encodeSubscribePayload subscribePacket
+payload (SUBACK subackPacket)           = encodeSubackPayload subackPacket
+payload (UNSUBSCRIBE unsubscribePacket) = encodeUnsubscribePaylooad unsubscribePacket
+payload _                               = mempty
+
+encodeUnsubscribePaylooad :: UnsubscribePacket -> Builder
+encodeUnsubscribePaylooad UnsubscribePacket{..} =
+  foldMap (encodeText . unTopicFilter) unsubscribeTopicFilters
+
+encodeSubackPayload :: SubackPacket -> Builder
+encodeSubackPayload SubackPacket{..} = foldMap (word8 . maybe 128 fromQoS) subackResponses
+
+encodeSubscribePayload :: SubscribePacket -> Builder
+encodeSubscribePayload SubscribePacket{..} =
+  foldMap
+    (\(topicFilter, qos) -> encodeText (unTopicFilter topicFilter) <> word8 (fromQoS qos))
+    subscribeTopicFiltersQoS
+
+encodePublishPayload :: PublishPacket -> Builder
+encodePublishPayload PublishPacket{..} = byteString $ messageMessage publishMessage
+
+encodeConnectPayload :: ConnectPacket -> Builder
+encodeConnectPayload ConnectPacket{..} =
+  mconcat
+    [ encodeText $ unClientIdentifier connectClientIdentifier
+    , maybe mempty (encodeText . unTopic . messageTopic) connectWillMsg
+    , maybe mempty (encodeBytes . messageMessage)        connectWillMsg
+    , maybe mempty (encodeText . unUserName)             connectUserName
+    , maybe mempty (encodeBytes . unPassword)            connectPassword
+    ]
+
+encodeRemaining :: Int64 -> Builder
```

Comment:
Better rename to `encodeRemainingLength`. Now it's not immediately clear encode remaining what?